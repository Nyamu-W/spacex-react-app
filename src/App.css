html {
  scroll-behavior: smooth;
}

.LaunchCard {
  margin-top: 20px !important;
}

.LaunchCardRow {
  margin: 20px 0;
}

.search {
  margin-bottom: 10px;
}

.crewImage {
  height: 398px;
  width: auto;
  object-fit: cover;
  object-position: top;
}

.shipImage {
  height: 398px;
  width: auto;
  object-fit: cover;
}

.fullPage {
  min-height: 90vh;
}

.footer {
  margin: 10px;
}

.shipDetailsLabel {
  font: 0.6em;
  font-weight: lighter;
  color: rgba(0, 0, 0, 0.7);
  margin-right: 10px;
}

.shipInfo {
  font-weight: bold;
  letter-spacing: 0.05em;
}
