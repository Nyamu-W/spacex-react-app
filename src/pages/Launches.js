import { Grid, Item, Pagination, Icon, Loader, Dropdown } from 'semantic-ui-react';
import { useState, useEffect } from 'react';
import LaunchCard from '../components/LaunchCard';

const dropdownOptions = [];
for (let i = new Date().getFullYear(); i >= 2006; i--) {
  dropdownOptions.push({
    key: i.toString(),
    value: i.toString(),
    text: i.toString(),
  })
}

const Launches = () => {
  const [launches, setLaunches] = useState([]);
  const [launchYears, setLaunchYears] = useState([]);
  const [currentPage, setCurrentPage] = useState(1);
  const [loading, setLoading] = useState(false);
  const [error, setError] = useState(false);

  const abortController = new AbortController();

  useEffect(() => {
    const getLaunches = async () => {
      const launchDataFromApi = await fetchLaunches();
      setLaunches(launchDataFromApi.reverse());
    }

    getLaunches();
    return () => abortController.abort();
  }, []);


  // Fetch launch data from API
  const fetchLaunches = async () => {
    setLoading(true);
    const data = await fetch('https://api.spacexdata.com/v4/launches/past', { signal: abortController.signal })
      .then(res => res.json())
      .catch(e => {
        setError(true);
        return []
      });
    setLoading(false);
    return data;
  }

  // Filter Launches
  const filteredLaunches = launches.filter(launch => {
    if (launchYears.length === 0) return true;
    return launchYears.some(year => launch.date_utc.includes(year));
  });

  // Pagination Variables
  const indexOfLastLaunch = currentPage * 5;
  const indexOfFirstLaunch = indexOfLastLaunch - 5;
  const currentLaunches = filteredLaunches.slice(indexOfFirstLaunch, indexOfLastLaunch);
  const totalPageCount = Math.ceil(filteredLaunches.length / 5);



  return (
    <>
      <Grid centered columns={1} padded >
        <Grid.Row />
        {error ?
          <h1>Error Loading. Try again later.</h1> :
          <>
            <Grid.Row>
              <h1>Launches</h1>
            </Grid.Row>
            <Grid.Row>
              <Dropdown
                onChange={(e, data) => setLaunchYears(data.value)}
                disabled={loading} placeholder='Filter by launch year'
                multiple
                selection
                options={dropdownOptions}
              />
            </Grid.Row>
            <Grid.Row stretched centered>
              {
                loading ?
                  <Loader size='massive' content='Loading Launch Data . . . ' active inline /> :
                  <Grid.Column
                    mobile='16'
                    tablet='16'
                    computer='14'
                    largeScreen='12'
                    widescreen='10'
                  >
                    <Item.Group divided link>
                      {currentLaunches.map(launch => <LaunchCard key={launch.id} launch={launch} />)}
                    </Item.Group>
                  </Grid.Column>
              }
            </Grid.Row>
            <Grid.Row>
              <Pagination
                activePage={currentPage}
                ellipsisItem={{ content: <Icon name='ellipsis horizontal' />, icon: true }}
                firstItem={{ content: <Icon name='angle double left' />, icon: true }}
                lastItem={{ content: <Icon name='angle double right' />, icon: true }}
                prevItem={{ content: <Icon name='angle left' />, icon: true }}
                nextItem={{ content: <Icon name='angle right' />, icon: true }}
                totalPages={totalPageCount}
                onPageChange={(e, data) => {
                  setCurrentPage(data.activePage);
                  window.scrollTo(0, 0);
                }}
              />
            </Grid.Row>
          </>
        }
      </Grid>
    </>
  )
}

export default Launches
