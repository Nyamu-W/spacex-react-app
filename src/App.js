import React, { } from 'react';
import { BrowserRouter as Router, Route } from 'react-router-dom';
import { Container } from 'semantic-ui-react';
import Launches from './pages/Launches';
import Ships from './pages/Ships';
import Crew from './pages/Crew';
import MenuBar from './components/MenuBar';
import SingleShip from './pages/SingleShip';
import SingleLaunch from './pages/SingleLaunch';
import SingleCrew from './pages/SingleCrew';
import Footer from './components/Footer';
import './App.css';

const App = () => {
  return (
    <Router>
      <Container fluid className='fullPage'>
        <MenuBar />
        <Route exact path="/" component={Launches} />
        <Route exact path="/launches" component={Launches} />
        <Route exact path="/launches/:launchId" component={SingleLaunch} />
        <Route exact path="/ships" component={Ships} />
        <Route exact path="/ships/:shipId" component={SingleShip} />
        <Route exact path="/crew" component={Crew} />
        <Route exact path="/crew/:crewId" component={SingleCrew} />
      </Container>
      <Footer />
    </Router>
  )
}

export default App
