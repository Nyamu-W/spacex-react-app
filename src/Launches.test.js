import { rest } from "msw";
import Launches from "./pages/Launches";
import renderer from "react-test-renderer";
import {
  render,
  findAllByText,
  waitForElementToBeRemoved,
  screen,
} from "@testing-library/react";
import App from "./App";
// import renderer  from 'react-test-renderer'

describe("Launches", () => {
  beforeEach(async () => {
    render(<App />);
    await waitForElementToBeRemoved(() => screen.getByText(/Loading/i));
  });

  it("displays launches", () => {
    expect(screen.getByText(/Test Launch 1/i)).toBeInTheDocument();
  });
});

it("Test test", () => {});
