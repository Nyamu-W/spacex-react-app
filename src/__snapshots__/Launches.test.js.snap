// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Launches fetch launches 1`] = `
Array [
  <div
    className="ui fluid container fullPage"
  >
    <div
      className="ui large inverted top fixed menu"
    >
      <div
        className="header item"
        onClick={[Function]}
      >
        SpaceX
      </div>
      <a
        className="active item"
        href="/launches"
        onClick={[Function]}
      >
        Launches
      </a>
      <a
        className="item"
        href="/ships"
        onClick={[Function]}
      >
        Ships
      </a>
      <a
        className="item"
        href="/crew"
        onClick={[Function]}
      >
        Crew
      </a>
    </div>
    <div
      className="ui centered padded one column grid"
    >
      <div
        className="row"
      />
      <div
        className="row"
      >
        <h1>
          Launches
        </h1>
      </div>
      <div
        className="row"
      >
        <div
          aria-disabled={false}
          aria-expanded={false}
          aria-multiselectable={true}
          className="ui multiple selection dropdown"
          onBlur={[Function]}
          onChange={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onMouseDown={[Function]}
          role="listbox"
          tabIndex={0}
        >
          <div
            aria-atomic={true}
            aria-live="polite"
            className="divider default text"
            role="alert"
          >
            Filter by launch year
          </div>
          <i
            aria-hidden="true"
            className="dropdown icon"
            onClick={[Function]}
          />
          <div
            className="menu transition"
          >
            <div
              aria-checked={false}
              aria-selected={true}
              className="selected item"
              onClick={[Function]}
              role="option"
              style={
                Object {
                  "pointerEvents": "all",
                }
              }
            >
              <span
                className="text"
              >
                2021
              </span>
            </div>
            <div
              aria-checked={false}
              aria-selected={false}
              className="item"
              onClick={[Function]}
              role="option"
              style={
                Object {
                  "pointerEvents": "all",
                }
              }
            >
              <span
                className="text"
              >
                2020
              </span>
            </div>
            <div
              aria-checked={false}
              aria-selected={false}
              className="item"
              onClick={[Function]}
              role="option"
              style={
                Object {
                  "pointerEvents": "all",
                }
              }
            >
              <span
                className="text"
              >
                2019
              </span>
            </div>
            <div
              aria-checked={false}
              aria-selected={false}
              className="item"
              onClick={[Function]}
              role="option"
              style={
                Object {
                  "pointerEvents": "all",
                }
              }
            >
              <span
                className="text"
              >
                2018
              </span>
            </div>
            <div
              aria-checked={false}
              aria-selected={false}
              className="item"
              onClick={[Function]}
              role="option"
              style={
                Object {
                  "pointerEvents": "all",
                }
              }
            >
              <span
                className="text"
              >
                2017
              </span>
            </div>
            <div
              aria-checked={false}
              aria-selected={false}
              className="item"
              onClick={[Function]}
              role="option"
              style={
                Object {
                  "pointerEvents": "all",
                }
              }
            >
              <span
                className="text"
              >
                2016
              </span>
            </div>
            <div
              aria-checked={false}
              aria-selected={false}
              className="item"
              onClick={[Function]}
              role="option"
              style={
                Object {
                  "pointerEvents": "all",
                }
              }
            >
              <span
                className="text"
              >
                2015
              </span>
            </div>
            <div
              aria-checked={false}
              aria-selected={false}
              className="item"
              onClick={[Function]}
              role="option"
              style={
                Object {
                  "pointerEvents": "all",
                }
              }
            >
              <span
                className="text"
              >
                2014
              </span>
            </div>
            <div
              aria-checked={false}
              aria-selected={false}
              className="item"
              onClick={[Function]}
              role="option"
              style={
                Object {
                  "pointerEvents": "all",
                }
              }
            >
              <span
                className="text"
              >
                2013
              </span>
            </div>
            <div
              aria-checked={false}
              aria-selected={false}
              className="item"
              onClick={[Function]}
              role="option"
              style={
                Object {
                  "pointerEvents": "all",
                }
              }
            >
              <span
                className="text"
              >
                2012
              </span>
            </div>
            <div
              aria-checked={false}
              aria-selected={false}
              className="item"
              onClick={[Function]}
              role="option"
              style={
                Object {
                  "pointerEvents": "all",
                }
              }
            >
              <span
                className="text"
              >
                2011
              </span>
            </div>
            <div
              aria-checked={false}
              aria-selected={false}
              className="item"
              onClick={[Function]}
              role="option"
              style={
                Object {
                  "pointerEvents": "all",
                }
              }
            >
              <span
                className="text"
              >
                2010
              </span>
            </div>
            <div
              aria-checked={false}
              aria-selected={false}
              className="item"
              onClick={[Function]}
              role="option"
              style={
                Object {
                  "pointerEvents": "all",
                }
              }
            >
              <span
                className="text"
              >
                2009
              </span>
            </div>
            <div
              aria-checked={false}
              aria-selected={false}
              className="item"
              onClick={[Function]}
              role="option"
              style={
                Object {
                  "pointerEvents": "all",
                }
              }
            >
              <span
                className="text"
              >
                2008
              </span>
            </div>
            <div
              aria-checked={false}
              aria-selected={false}
              className="item"
              onClick={[Function]}
              role="option"
              style={
                Object {
                  "pointerEvents": "all",
                }
              }
            >
              <span
                className="text"
              >
                2007
              </span>
            </div>
            <div
              aria-checked={false}
              aria-selected={false}
              className="item"
              onClick={[Function]}
              role="option"
              style={
                Object {
                  "pointerEvents": "all",
                }
              }
            >
              <span
                className="text"
              >
                2006
              </span>
            </div>
          </div>
        </div>
      </div>
      <div
        className="centered stretched row"
      >
        <div
          className="fourteen wide computer twelve wide large screen sixteen wide mobile sixteen wide tablet ten wide widescreen column"
        >
          <div
            className="ui divided link items"
          />
        </div>
      </div>
      <div
        className="row"
      >
        <div
          aria-label="Pagination Navigation"
          className="ui pagination menu"
          role="navigation"
        >
          <a
            aria-current={false}
            aria-disabled={false}
            className="icon item"
            onClick={[Function]}
            onKeyDown={[Function]}
            tabIndex={0}
            type="firstItem"
            value={1}
          >
            <i
              aria-hidden="true"
              className="angle double left icon"
              onClick={[Function]}
            />
          </a>
          <a
            aria-current={false}
            aria-disabled={false}
            className="icon item"
            onClick={[Function]}
            onKeyDown={[Function]}
            tabIndex={0}
            type="prevItem"
            value={1}
          >
            <i
              aria-hidden="true"
              className="angle left icon"
              onClick={[Function]}
            />
          </a>
          <a
            aria-current={false}
            aria-disabled={false}
            className="icon item"
            onClick={[Function]}
            onKeyDown={[Function]}
            tabIndex={0}
            type="nextItem"
            value={0}
          >
            <i
              aria-hidden="true"
              className="angle right icon"
              onClick={[Function]}
            />
          </a>
          <a
            aria-current={false}
            aria-disabled={false}
            className="icon item"
            onClick={[Function]}
            onKeyDown={[Function]}
            tabIndex={0}
            type="lastItem"
            value={0}
          >
            <i
              aria-hidden="true"
              className="angle double right icon"
              onClick={[Function]}
            />
          </a>
        </div>
      </div>
    </div>
  </div>,
  <div
    className="footer"
  >
    <div
      className="ui labeled button right floated"
      onClick={[Function]}
    >
      <button
        className="ui black mini basic compact button"
        tabIndex={0}
      >
         
        Data Source
      </button>
      <a
        className="ui black left pointing label"
        href="https://github.com/r-spacex/SpaceX-API"
        onClick={[Function]}
        target="blank"
      >
        <i
          aria-hidden="true"
          className="rocket icon"
          onClick={[Function]}
        />
        SpaceX API
      </a>
    </div>
  </div>,
]
`;
